File: 0011-container-with-most-water/0011-container-with-most-water.cpp
// @Author: KING-SEN
// 1 last try

class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0, right = height.size() - 1;
        int answer = 0;
        while (left <= right) {
            int area = min(height[left], height[right]) * (right - left);
            answer = max(answer, area);
            if (height[left] > height[right]) {
                right--;
            }
            else {
                left++;
            }
        }
        return answer;
    }
};

// Time Complexity: O(N), Where N is length of the array
// Space Complexity: O(1), It uses constant space

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        answer = 0
        while l < r:
            answer = max(answer, min(height[l], height[r]) * (r - l))
            if height[l] <= height[r]:
                l += 1
            else:
                r -= 1
        return answer
'''
Time complexity: O(n)
Space complexity: O(1)
'''